### PR TITLE
Fixing a couple of held locks during exception handling

### LINKED ISSUE #2959: Program terminates during error handling
While debugging the following code snippet
``` C++
#include <hpx/hpx_main.hpp>

#include <hpx/include/thread_executors.hpp>

int main()
{
    hpx::threads::executors::local_priority_queue_os_executor exec(1);

    return 0;
}
```
I encountered what seems to be a bug in the error handling path. The program aborts with the error
```
terminate called after throwing an instance of 'hpx::detail::exception_with_info<hpx::exception>'
  what():  suspending thread while at least one lock is being held, stack backtrace: 25 frames:
```
during unwinding.
The emitted stack trace is:
```
0x7ffff6cf92a4  : hpx::detail::backtrace_direct[abi:cxx11](unsigned long) + 0xb4 in /opt/hpx-debug/lib/libhpxd.so.1
0x7ffff741caf1  : hpx::util::verify_no_locks() + 0xf1 in /opt/hpx-debug/lib/libhpxd.so.1
0x7ffff7299096  : hpx::this_thread::suspend(hpx::threads::thread_state_enum, boost::intrusive_ptr<hpx::threads::thread_data> const&, hpx::util::thread_description const&, hpx::error_code&) + 0xa6 in /opt/hpx-debug/lib/libhpxd.so.1
0x7ffff7290daf  : hpx::this_thread::yield_to(hpx::thread::id) + 0x3f in /opt/hpx-debug/lib/libhpxd.so.1
0x7ffff73ab541  : hpx::util::backtrace::trace_on_new_stack[abi:cxx11]() const + 0x1e1 in /opt/hpx-debug/lib/libhpxd.so.1
0x7ffff6cf90fa  : hpx::detail::backtrace[abi:cxx11](unsigned long) + 0x5a in /opt/hpx-debug/lib/libhpxd.so.1
0x7ffff6d0e56d  : std::__exception_ptr::exception_ptr hpx::detail::get_exception<hpx::exception>(hpx::exception const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, long, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xbd in /opt/hpx-debug/lib/libhpxd.so.1
0x7ffff6d0eafc  : void hpx::detail::throw_exception<hpx::exception>(hpx::exception const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, long) + 0x6c in /opt/hpx-debug/lib/libhpxd.so.1
0x7ffff6d67340  : hpx::detail::throw_exception(hpx::error, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, long) + 0x60 in /opt/hpx-debug/lib/libhpxd.so.1
0x7ffff71e64a9  : ??? + 0x7ffff71e64a9 in /opt/hpx-debug/lib/libhpxd.so.1
0x7ffff71e7061  : hpx::resource::detail::partitioner::get_pool_data(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x221 in /opt/hpx-debug/lib/libhpxd.so.1
0x7ffff71e77a1  : hpx::resource::detail::partitioner::assign_pu(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long) + 0x101 in /opt/hpx-debug/lib/libhpxd.so.1
0x7ffff72127d5  : hpx::threads::detail::scheduled_thread_pool<hpx::threads::policies::local_priority_queue_scheduler<std::mutex, hpx::threads::policies::lockfree_fifo, hpx::threads::policies::lockfree_fifo, hpx::threads::policies::lockfree_lifo> >::add_processing_unit(unsigned long, unsigned long, std::shared_ptr<hpx::compat::barrier>, hpx::error_code&) + 0x55 in /opt/hpx-debug/lib/libhpxd.so.1
0x7ffff72270f0  : hpx::threads::detail::scheduled_thread_pool<hpx::threads::policies::local_priority_queue_scheduler<std::mutex, hpx::threads::policies::lockfree_fifo, hpx::threads::policies::lockfree_fifo, hpx::threads::policies::lockfree_lifo> >::run(std::unique_lock<std::mutex>&, unsigned long) + 0x4d0 in /opt/hpx-debug/lib/libhpxd.so.1
0x7ffff7276036  : ??? + 0x7ffff7276036 in /opt/hpx-debug/lib/libhpxd.so.1
0x7ffff7272d6f  : hpx::threads::executors::local_priority_queue_os_executor::local_priority_queue_os_executor(unsigned long, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x2f in /opt/hpx-debug/lib/libhpxd.so.1
0x555555557fb1  : ??? + 0x555555557fb1 in /home/christopher/test/hpx/thread_pool_executor_test/cmake-build-debug/thread_pool_executor_test
0x55555555b850  : ??? + 0x55555555b850 in /home/christopher/test/hpx/thread_pool_executor_test/cmake-build-debug/thread_pool_executor_test
0x55555555b6b7  : ??? + 0x55555555b6b7 in /home/christopher/test/hpx/thread_pool_executor_test/cmake-build-debug/thread_pool_executor_test
0x55555555b168  : ??? + 0x55555555b168 in /home/christopher/test/hpx/thread_pool_executor_test/cmake-build-debug/thread_pool_executor_test
0x7ffff6d5234f  : hpx::runtime_impl::run_helper(hpx::util::function<int (), false>, int&) + 0x3af in /opt/hpx-debug/lib/libhpxd.so.1
0x7ffff6d531cc  : ??? + 0x7ffff6d531cc in /opt/hpx-debug/lib/libhpxd.so.1
0x7ffff71f3d28  : hpx::threads::coroutines::detail::coroutine_impl::operator()() + 0xc8 in /opt/hpx-debug/lib/libhpxd.so.1
0x7ffff71f6c7a  : ??? + 0x7ffff71f6c7a in /opt/hpx-debug/lib/libhpxd.so.1: HPX(invalid_status)
```

Since the original issue does not seem to be related to this bug, I will report it separately.

### PR DESCRIPTION
**Pull request by hkaiser**
- making sure exceptions thrown during construction of a local_priority_queue_os_executor get propagated properly
- this fixes #2959 

@chinz07 please verify whether this fixes your issue

### CODE REVIEW

**Comment by chinz07** [comment]
@hkaiser This fixes the issue, thanks!

File: hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp [hunk 1]
```diff
@@ -292,7 +292,7 @@ namespace hpx { namespace threads { namespace detail
                          << " failed with: " << e.what();
 
             // trigger the barrier
-            pool_threads -= thread_num;
+            pool_threads -= (thread_num + 1);
```
Comment (sithhell) [inline hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp]: Why did you make this change?

Comment (hkaiser) [reply hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp]: Without this change the loop below tries to wait for one too many thread. `thread_num` gets incremented only after the thread has successfully started, so any exception thrown during thread startup will see `thread_num` not reflecting the current (failed) thread.

Comment (sithhell) [reply hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp]: Shouldn't it be rather `--pool_threads` then? `thread_num + 1` itself might decrease the value too much, no?

Comment (hkaiser) [reply hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp]: what's the difference between
```
    pool_threads -= (thread_num + 1); 
```
and
```
    --pool_threads;
    pool_threads -= thread_num; 
```
?

Comment (sithhell) [reply hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp]: I meant just `--pool_threads`, not both.

Comment (hkaiser) [reply hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp]: That will not work.

Comment (sithhell) [reply hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp]: Ok, missed that the catch block is covering the loop as well.

Comment (hkaiser) [reply hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp]: I take that as an approval of this PR...

**Pull request merged**